Support reading from S3 for AHI-HSD

Isotr0py opened the pull request:
- Add support for reading AHI-HSD data from AWS S3

Since NOAA provides open AHI-HSD data on [noaa-himawari](https://noaa-himawari9.s3.amazonaws.com/index.html), I think it's possible to read remote Himawari data from AWS S3 with just some simple changes on `satpy/readers/util.py`.

Here is my simple test code:
```python
# a simple test code
from satpy import Scene

reader_kwargs = {
    'storage_options': {
        's3': {'anon': True},
        'simple': {
            'cache_storage': '/tmp/s3_cache',
        }
    }
}

filenames = ['simplecache::s3://noaa-himawari9/AHI-L1b-FLDK/2023/03/24/0450/HS_H09_20230324_0450_B0*']
scn = Scene(filenames=filenames, reader='ahi_hsd',reader_kwargs=reader_kwargs)
scn.load(['true_color'])
new_scn = scn.resample(scn.coarsest_area(), resampler='native')
new_scn.save_dataset('true_color', filename='himawari_ahi_truecolor_03240450.png')
```

And the output should be like this:
![himawari_ahi_truecolor_03240450](https://user-images.githubusercontent.com/41363108/227448160-91170b27-9eda-43a5-ac6d-3afddb2a67da.png)
(I resize the output image because the original image is too large to show)

 - [ ] Fully documented
 - [x] Add your name to `AUTHORS.md` if not there already

djhoese commented on satpy/readers/utils.py:
```diff
@@ -210,56 +211,71 @@ def unzip_file(filename, prefix=None):
-                              % (filename, status))
+    fdn, tmpfilepath = tempfile.mkstemp(prefix=prefix,
+                                                dir=config["tmp_dir"])
+    if type(filename) == str:
```
```suggestion
    if isinstance(filename, str):
```

djhoese reviewed (changes requested):
Thanks for starting the work on this. This will be great to have.

Could you try addressing some of the issues identified by the `pre-commit.ci` check? Things like over-indenting or under-indenting things. Otherwise, the main things left to do on this PR:

1. Add a unit test
2. Cleanup the `unzip_file` function by splitting it up into separate functions to be easier to read.

Until then it is a little difficult to decipher how this is working.

Isotr0py commented:
I split `unzip_FSFile` from `unzip_file` function to load remote `.bz2` file respectively.
I also added a unit test to test `unzip_FSFile`. However, I'm not very familiar with `unittest`, so I'm not sure whether there are any tests left out.

djhoese commented on satpy/readers/ahi_hsd.py:
```diff
@@ -357,7 +359,10 @@ def __init__(self, filename, filename_info, filetype_info,
 
         self.is_zipped = False
-        self._unzipped = unzip_file(self.filename, prefix=str(filename_info['segment']).zfill(2))
+        if isinstance(self.filename, str):
+            self._unzipped = unzip_file(self.filename, prefix=str(filename_info['segment']).zfill(2))
+        elif isinstance(self.filename, FSFile):
+            self._unzipped = unzip_FSFile(self.filename, prefix=str(filename_info['segment']).zfill(2))
```
Could you put this `if/elif` into a `unzip_file` function and then refactor that function to use a `_unzip_local_file` and an `_unzip_fsfile`? There is a lot of duplicate code here so it'd be nice to hide it away in the helper/utility function.

djhoese commented on satpy/readers/utils.py:
```diff
@@ -263,6 +264,38 @@ def unzip_file(filename, prefix=None):
+        Temporary filename path for decompressed file or None.
+
+    """
+    if str(filename).endswith('bz2'):
```
You could flip this logic around and do:

```
if not str(filename).endswith('bz2'):
    return None
... other code ...
```

djhoese commented on satpy/tests/reader_tests/test_utils.py:
```diff
@@ -344,6 +344,36 @@ def test_generic_open_filename(self, open_mock):
+        segmentstr = str(segment).zfill(2)
+
+        new_fname = hf.unzip_FSFile(fsf, prefix=segmentstr)
+        assert mock_bz2_decompress.called
```
The mock object has a `assert_called` method you could use without the `assert ` statement before it.

djhoese reviewed (changes requested):
I have a couple more requests, but this is looking better.

@pnuu @simonrp84 @mraspaud as the people who I think have dealt with remote reading the most and Simon I think with bz2 stuff, could you take a look at this?

mraspaud commented:
Ok, so I looked into this a bit, and I can see that for FSFile instances, the file is decompressed in a temporary file given by `mkstemp`. I this this is fine, but we have the possibility to shortcircuit this, and remove a bit of code from this PR.

So, in the example above, if you add `"compression": "bz2"` in the storage options, fsspec will do bz2 decompression on the fly. That means that the unzipping need to be deactivated for FSFiles, but the rest could be almosst identical...

...except the `np.fromfile` calls, that expect a file open from disk, not a higher level File object. An alternative is to make our own `fromfile` function, that would look like this:
```python
def fromfile(fd, dtype=None, count=None, **kwargs):
	data = fd.read(dtype.itemsize * count)
	return np.frombuffer(data, dtype=dtype, count=count, **kwargs)
```

Now, this should make things easier to follow, but the real question is how much does this affect performance in the end...

djhoese commented on satpy/readers/utils.py:
```diff
@@ -263,6 +264,38 @@ def unzip_file(filename, prefix=None):
+                traceback.print_exc()
+                LOGGER.info("Failed to read bzipped file %s", str(filename))
+                os.remove(tmpfilepath)
+                return None
```
Ah I found it. It was @adybbroe a long time ago in this commit: https://github.com/pytroll/satpy/commit/8a1cdeb781496654ed1d23ad67b546fe7472868a#diff-94d5228d6fa7ede1b6aea206da7c1426f32e2c133f4768233b42527363a46898

I think we could change this to a `LOGGER.debug` as I mentioned above. @adybbroe what do you think?

adybbroe replied:
Yes, I do remember vaguely now. There should be no problem changing to `LOGGER.debug` from my point of view. Go ahead!

mraspaud commented:
So I tried what I suggested above and, factoring out the download time using simplecache, I get roughly the same performance as working with files on disk.

Isotr0py commented:
@mraspaud I got it! 
However, it seems that other readers share the `unzip_file` function with `ahi_hsd` reader. So if `unzip_file` return bytes directly instead of a temporary file, there may be some issues for other readers.

Anyway, I think it's possible to check the file type from their header to skip unzipping or not in `unzip_file`, and it will make reading FSFile more flexibly.

Isotr0py commented:
I add the code to identify data stream to activate/skip unzip:
```python
if content.startswith(bytes.fromhex("425A68")):
    content = bz2.decompress(content)
```
Now if `"compression": "bz2"` added in the storage options, unzipping will be skipped for FSFile, else unzipping will activate locally.

djhoese commented on satpy/tests/reader_tests/test_utils.py:
```diff
@@ -344,6 +344,43 @@ def test_generic_open_filename(self, open_mock):
+        new_fname = hf.unzip_file(fsf, prefix=segmentstr)
+        mock_bz2_decompress.assert_not_called
+        bz2_mock.assert_not_called
+        assert os.path.exists(new_fname) is True
```
```suggestion
        assert os.path.exists(new_fname)
```

djhoese commented on satpy/tests/reader_tests/test_utils.py:
```diff
@@ -344,6 +344,43 @@ def test_generic_open_filename(self, open_mock):
+        new_fname = hf.unzip_file(fsf, prefix=segmentstr)
+        mock_bz2_decompress.assert_called
+        bz2_mock.assert_called
+        assert os.path.exists(new_fname) is True
```
```suggestion
        assert os.path.exists(new_fname)
```

djhoese commented on satpy/readers/utils.py:
```diff
@@ -210,57 +230,84 @@ def unzip_file(filename, prefix=None):
-
-        # Otherwise, fall back to the original method
-        bz2file = bz2.BZ2File(filename)
+    if not os.fspath(filename).endswith('bz2'):
+        return None
```
Can `os.fspath` be used for both FSFile objects and strings? If so, then could this if statement be moved to the parent function and removed from each of these `_unzip_X` functions?

djhoese commented on satpy/readers/utils.py:
```diff
@@ -210,57 +230,84 @@ def unzip_file(filename, prefix=None):
+    if content.startswith(bytes.fromhex("425A68")):
+        content = bz2.decompress(content)
+
+    with closing(os.fdopen(fdn, 'wb')) as ofpt:
+        try:
+            ofpt.write(content)
+        except IOError:
+            LOGGER.debug("Failed to read bzipped file %s", str(filename))
+            os.remove(tmpfilepath)
+            return None
+    return tmpfilepath
```
This block of code is shared between the two helper functions. Could it be extracted and the caller just provides the `content` bytes? Oh I guess it would also need `fdn`, `filename`, and `tmpfilepath`. That's unfortunate, but I think the cleaner and less duplicated code is worth it. I think it could be called like:

```python
return _write_uncompressed_bz2_data(content, fdn, filename, tmpfilepath)
```

djhoese replied:
Side note: Would the memory saving be worth it to decompress blocks of data at a time? Basically open the destination file, read N bytes of data, decompress them, write the decompressed data, then loop again and read N bytes of compressed data, decompress, and so on.

@simonrp84 thoughts? Maybe in another PR?

djhoese commented on satpy/readers/utils.py:
```diff
@@ -210,57 +230,84 @@ def unzip_file(filename, prefix=None):
+    # try pbzip2
+    pbzip = which('pbzip2')
+    # Run external pbzip2
+    if pbzip is not None:
```
Is it possible to put this entire `pbzip/which` and `if` statement in a separate helper function like `_decompress_with_pbzip_if_available`? If it returns the tmpfilepath then we know it was successful and we can return `tmpfilepath`. If it is `None` then we know `pbzip` wasn't available and we should continue with the `bz2` module decompression.

djhoese reviewed (changes requested):
Looking good. I had a lot more refactoring questions. Sorry for how many there are. I should have bugged the original authors of this code with these things when they were first added. I do think this code is getting much cleaner which is nice to see. Thanks for sticking with this.

djhoese commented on satpy/readers/utils.py:
```diff
@@ -210,57 +234,85 @@ def unzip_file(filename, prefix=None):
+        content = bz2.decompress(content)
+
+    return _write_uncompress_file(content, fdn, filename, tmpfilepath)
+# endregion
```
```suggestion
```

Please remove the start and end "region" comments.

djhoese reviewed (changes requested):
One small small request. Otherwise looks good. The code is so much cleaner than it was before you started this PR. Thanks for working on this.

mraspaud commented on satpy/readers/utils.py:
```diff
@@ -210,57 +233,84 @@ def unzip_file(filename, prefix=None):
+    with closing(os.fdopen(fdn, 'wb')) as ofpt:
+        try:
+            ofpt.write(content)
+        except IOError:
+            LOGGER.debug("Failed to read bzipped file %s", str(filename))
+            os.remove(tmpfilepath)
+            return None
```
This should be place back around the `.read` call, as I think the IOError we want to catch if for reading, not writing.

Isotr0py replied:
Maybe adding IOError catch for both reading and writing will be better?

Isotr0py replied:
Maybe adding IOError catch for both reading and writing will be better?

mraspaud replied:
Yes, that works for me :)

Isotr0py replied:
OK, leaving them unhandled is also an appropriate solution :)

mraspaud commented on AUTHORS.md:
```diff
@@ -82,3 +82,4 @@ The following people have made contributions to this project:
 - [praerien (praerien)](https://github.com/praerien)
 - [Xin Zhang (zxdawn)](https://github.com/zxdawn)
 - [Yufei Zhu (yufeizhu600)](https://github.com/yufeizhu600)
+- [Zifeng Mo (Isotr0py)](https://github.com/Isotr0py)
```
We are sorting the names by alphabetical order of last names. So this is correct, right?

mraspaud reviewed (commented):
Thanks a lot for this contribution, it will allow adapting readers to FSFiles easier :) I have a few comments inline, but otherwise I think this is almost ready to be merged!

sfinkens commented on satpy/readers/utils.py:
```diff
@@ -210,57 +233,84 @@ def unzip_file(filename, prefix=None):
+    if pbzip2 is not None:
+        return pbzip2
+    # Otherwise, fall back to the original method
+    bz2file = bz2.BZ2File(filename)
+    content = bz2file.read()
```
For better readability, this could be extracted to a separate method `_unzip_with_bz2file`

Isotr0py replied:
OK, I have extracted this with `_unzip_with_bz2file` and added IOerror catch for file reading.

sfinkens commented on satpy/readers/utils.py:
```diff
@@ -210,57 +233,84 @@ def unzip_file(filename, prefix=None):
+    return tmpfilepath
+
+
+def _write_uncompress_file(content, fdn, filename, tmpfilepath):
```
Is this a typo? uncompress -> uncompress**ed**

sfinkens reviewed (commented):
Nice work, thanks a lot for your contribution! Just two minor comments

mraspaud commented:
Looks like the tests are failing now...

Isotr0py commented:
It seems that using `with bz2.BZ2File(filename) as bz2file:` will cause the unit test failed...

djhoese commented:
I could be wrong, but this seems like the wrong fix for the tests. The tests are failing because the mock didn't handle the context manager usage (`with` statement). Right? The tests pass now, but at the cost of the regular code being messier.

Isotr0py commented:
@djhoese Thanks for your tips! I just adapted `bz2_mock` to the context manager usage, and the test using `with` statement should pass.
